File: lbb/2017-11-12-3Sum.cpp
/*
Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0? Find all unique triplets in the array which gives the sum of zero.

Note: The solution set must not contain duplicate triplets.

For example, given array S = [-1, 0, 1, 2, -1, -4],

A solution set is:
[
  [-1, 0, 1],
  [-1, -1, 2]
]

题目链接 https://leetcode.com/problems/3sum/#/description
*/

public class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        int length = nums.length;
        List<List<Integer>> res = new LinkedList<List<Integer>>();
        List<Integer> tmp = new LinkedList<Integer>();
        
        for(int i = 0; i < length-1; i++){
            int target = -nums[i];
            int front = i+1;
            int back = length - 1;
            while(front<back){
                int sum = nums[front]+nums[back];  
                if(sum == target){
                    tmp.add(nums[i]);
                    tmp.add(nums[front]);                    
                    tmp.add(nums[back]);
                    res.add(tmp);
                    tmp = new LinkedList<Integer>();
                    while(front<back && nums[front]==nums[front+1]) front++;                
                    while(front<back && nums[back]==nums[back-1]) back--;
                    front++;
                    back--;
                }else if( sum < target){
                    front++;
                }
                else{
                    back--;
                }
            }
            while( (i<length-1) && nums[i]==nums[i+1]) i++;
        }
        return res;        
    }
}

File: lbb/2017-11-12-3Sum.js
//https://leetcode.com/problems/3sum/description/
var arrSort = function(arr){
    arr.sort(function(a,b){
        if(a>b) return 1;
        else if(a<b) return -1;
        return 0;
    })
    return arr;
}
function threeSum(nums){
    var arr2 = arrSort(nums);
    // let arr2 = Array.from(new Set(arr1))
    var length = arr2.length;
    let res = [];
    for(let index = 0; index<length; index++){
        let sum = 0 - arr2[index];
        let element = arr2[index];
        let i = index + 1, j = arr2.length - 1;
        while(i<j){
            if(arr2[i] + arr2[j] == sum){
                res.push([element, arr2[i],arr2[j]]);
                while( i<j & arr2[i] == arr2[i+1]) i++;
                while( i<j & arr2[j] == arr2[j+1]) j--;
                i++;
                j--;
            }
            while(arr2[i] + arr2[j] < sum) i++;
            while(arr2[i] + arr2[j] > sum) j--;
        }
        while(index<length & arr2[index]==arr2[index+1]) index++;
    }
    return res;
}
console.log(threeSum([-1, 0,3,-7,-2,0,1,-1,-1,2]))